Automattic/jetpack
# Launchpad Checklist API: Add checklist and task registry

Fixes #

## Proposed changes:

* Create a centralized registry for Launchpad checklists and Launchpad checklist tasks
* By doing so, we can create easy to use helper methods ( `register_launchpad_task_list`, `register_launchpad_task` ) that will allow other developers to spin up Launchpad checklists quickly
* To elaborate, all of this will allow others to programmatically register Launchpad checklists and tasks wherever is convenient in wpcom rather than hardcoding information in the `jetpack-mu-wpcom` `launchpad.php` file

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:






1. Run `bin/jetpack-downloader test jetpack-mu-wpcom-plugin add/registry-for-launchpad-checklists` to build and sync this branch to your sandbox. 
2. Sandbox a test site and public API
3. Start any new launchpad onboarding site Ex. wordpress.com/setup/free/intro
4. Go to `https://developer.wordpress.com/docs/api/console/`, select Rest API and wpcom/v2, and input the following: `/sites/YOURSITESLUG/launchpad/checklist?checklist_slug=free`. Confirm you get back an appropriate checklist that matches the status of checklist items on Launchpad. 
5. Add the following code to the bottom of the `launchpad.php` file. We are going to verify that the checklist registration logic works properly by programmatically adding an entirely new checklist

```
function register_fake_checklist_and_tasks() {
	$fake_checklist = array(
		'id'          => 'fake_checklist',
		'title'       => 'Fake Checklist',
		'task_ids'       => array(
			'fake_task_1',
			'fake_task_2',
		),
	);

	register_launchpad_task_list( $fake_checklist );

	$fake_tasks = array(
		array(
			'id'        => 'fake_task_1',
			'title'     => 'Fake Task 1',
			'completed' => false,
			'disabled'  => false,
		),
		array(
			'id'        => 'fake_task_2',
			'title'     => 'Fake Task 2',
			'completed' => false,
			'disabled'  => false,
		)
		);

	register_launchpad_tasks( $fake_tasks );
}

add_action( 'init', 'register_fake_checklist_and_tasks' );
```
6. Then add `fake_checklist` as [a whitelisted query param on line 53](https://github.com/Automattic/jetpack/blob/trunk/projects/packages/jetpack-mu-wpcom/src/features/wpcom-endpoints/class-wpcom-rest-api-v2-endpoint-launchpad-checklist.php#L53) of the `class-wpcom-rest-api-v2-endpoint-launchpad-checklist.php` file
7. Verify in the developer console that `/sites/YOURSITESLUG/launchpad/checklist?checklist_slug=fake_checklist` returns a collection of fake tasks that we registered. 

<img width="844" alt="Screenshot 2023-04-20 at 7 01 18 PM" src="https://user-images.githubusercontent.com/5414230/233523239-15094610-5339-44b2-9022-8057adbcde50.png">
